scipy/linalg/_matfuncs_inv_ssq.py

```diff
@@ -760,13 +760,20 @@ def _logm_triu(T):
 
     # Construct T0 with the appropriate type,
     # depending on the dtype and the spectrum of T.
-    T_diag = np.diag(T)
+    T_diag = T.diagonal()
     keep_it_real = np.isrealobj(T) and np.min(T_diag) >= 0
     if keep_it_real:
         T0 = T
     else:
         T0 = T.astype(complex)
 
+    # Check if T0 is diagonal, if so return logarithm of the diagonal
+    # Reshape it so that diagonal becomes first columnn
+    T0_diagonal_reshape = T0.reshape(-1)[:-1].reshape(n-1,n+1)
+    is_diagonal = np.allclose(T0_diagonal_reshape[:,1:], 0)
```

Comment:
allclose has atol just 1e-8. The off-diagonal elements probably should be compared to the diagonal ones (which of them?) to determine whether the matrix is purely diagonal or not.

Reply:
Taking a peek at Higham section 11.2 about the conditioning of the matrix logarithm, we see that for positive definite matrices the conditioning number of log(A) is k(A)/||log(A)|| (and for non-PD this is a lower bound). Therefore if the off diagonal elements are small compared to the smallest diagonal element, it should be stable. (but this is not really a formal proof either)

Perhaps we can check whether the ratio of smallest diagonal element to biggest off-diagonal is bounded by e.g. 1e-8.

I can try to get a formal proof going

Reply:
I did some numerical experiments. We take a diagonal D matrix of norm 1, with entries following a linear scale (to create an arbitrarily high condition number). We perturb it with a uniform random upper triangular matrix P. 
We then record the relative error between logm(D) and logm(D+P) as function of the ratio between the minimal entry of D and maximal entry of P. The logarithms of the relative errors and this min/max ratio seem to follow a linear relationship.
In practice if this ratio is ~1e-8, then the relative error is also around 1e-8 (and in fact lower if the condition number is very high)

Therefore it seems that this ratio of smallest diagonal to biggest off-diagonal is a good indicator. 

Then we just have to fix a good value for this ratio. The goal is that the method is faster for PD matrices, so let's take the Schur decomposition of some PD matrices and record what this ratio is for the upper triangular matrix of the decomposition. It seems that the log of this ratio depends linearly on the log of the condition number of the PD matrix. 
At a condition number of ~1e7 the ratio is around 1e-7 to 1e-9. 

Even at very high condition number (>1e10), the we still see that approximation using only the diagonal is good; it both corresponds well to scipy's matrix logarithm, and if we exponentiate it we recover the matrix to a relative error of <1e-14. 

All in all, it seems that a ratio of 1e-8 is a reasonable tradeoff.